metrics_sdk/lib/opentelemetry/sdk/metrics/state/metric_stream.rb

```diff
@@ -0,0 +1,80 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module SDK
+    module Metrics
+      module State
+        # @api private
+        #
+        # The MetricStream class provides SDK internal functionality that is not a part of the
+        # public API.
+        class MetricStream
+          attr_reader :name, :description, :unit, :instrument_kind, :instrumentation_library, :data_points
+
+          def initialize(
+            name,
+            description,
+            unit,
+            instrument_kind,
+            meter_provider,
+            instrumentation_library
+          )
+            @name = name
+            @description = description
+            @unit = unit
+            @instrument_kind = instrument_kind
+            @meter_provider = meter_provider
+            @instrumentation_library = instrumentation_library
+
+            @data_points = {}
+            @mutex = Mutex.new
+          end
+
+          def collect(start_time, end_time)
```

Comment:
Does `data_points` keep growing without bound? I'm not clear on the expected behaviour here - part of me thinks we're supposed to only return data points recorded during the interval `start_time .. end_time`, but I'm really not sure.

Reply:
That needs to be addressed as part of the aggregation temporality work that is not part of this PR, i.e. sum or delta.

In the case of sum, there should be some concept of pruning. It's mentioned in the [guidelines](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/supplementary-guidelines.md#guidelines-for-sdk-authors).

>So here are some suggestions that we encourage SDK implementers to consider:
> * You want to control the memory usage rather than allow it to grow indefinitely / unbounded - regardless of what aggregation temporality is being used.
> * You want to improve the memory efficiency by being able to forget about things that are no longer needed.
> * You probably don't want to keep exporting the same thing over and over again, if there is no updates. You might want to consider [Resets and Gaps](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/data-model.md#resets-and-gaps). For example, if a Cumulative metrics stream hasn't received any updates for a long period of time, would it be okay to reset the start time?